### ISSUE TITLE
Documentation: services.netbox.plugins

### ISSUE THREAD

**Issue by brotherdust** [NixOS/nixpkgs#261522]
## Problem

The documentation for this attribute is inadequate. Here's the definition:
```
 plugins = lib.mkOption {
      type = with lib.types; functionTo (listOf package);
      default = _: [];
      defaultText = lib.literalExpression ''
        python3Packages: with python3Packages; [];
      '';
      description = lib.mdDoc ''
        List of plugin packages to install.
      '';
    };
```

Okay... what should this actually look like, then? Does it require that the python package must be present in nixpkgs? What if it's not? Can examples be provided to make those plugins available?

## Proposal

1. Enhance the documentation to include an example of what this option will accept. The current value of `defaultText` is too obtuse. How would a newcomer know how to interpret this?
2. If the Python package is not available in nixpkgs, provide a link to or in-line example of how to make the package available. I.E. `netbox-bgp`

Thank you kindly!

## Checklist

- [x] checked [latest Nixpkgs manual] \([source][nixpkgs-source]) and [latest NixOS manual] \([source][nixos-source])
- [x] checked [open documentation issues] for possible duplicates
- [x] checked [open documentation pull requests] for possible solutions

[latest Nixpkgs manual]: https://nixos.org/manual/nixpkgs/unstable/
[latest NixOS manual]: https://nixos.org/manual/nixos/unstable/
[nixpkgs-source]: https://github.com/NixOS/nixpkgs/tree/master/doc
[nixos-source]: https://github.com/NixOS/nixpkgs/tree/master/nixos/doc/manual
[open documentation issues]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+label%3A%229.needs%3A+documentation%22
[open documentation pull requests]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+is%3Apr+label%3A%228.has%3A+documentation%22%2C%226.topic%3A+documentation%22

**Comment by felbinger** [comment]
@brotherdust Did you find a solution to install a plugin?
I'm trying to setup `peering-manager` with oidc which requires `mozilla-django-oidc` - it's the same schema there, because it's also a django app.

**Comment by felbinger** [comment]
> @brotherdust Did you find a solution to install a plugin? I'm trying to setup `peering-manager` with oidc which requires `mozilla-django-oidc` - it's the same schema there, because it's also a django app.

I found a solution to my problem, in case someone else needs an example:
```nix
{ config
, lib
, pkgs
, ...
}: let 
  mozilla-django-oidc = ps: ps.buildPythonPackage rec {
    pname = "mozilla-django-oidc";
    version = "2.0.0";
    format = "pyproject";
    src = ps.fetchPypi {
      inherit pname version;
      hash = "sha256-qLLyfGnBItL02AHDdZdh0z3r8Grp2ruriu2CiHu6O7g";
    };
    nativeBuildInputs = with ps; [
      setuptools
    ];
    propagatedBuildInputs = with ps; [
      django
      josepy
      requests
      cryptography
    ];
    meta = with lib; {
      description = "A django OpenID Connect library";
      homepage = "https://github.com/mozilla/mozilla-django-oidc";
      license = licenses.mpl20;
      platforms = platforms.linux;
    };
  };
in  {
  sops.secrets = {
    "peering-manager/secretKey".owner = "peering-manager";
    "peering-manager/oidcSecret".owner = "peering-manager";
  };

  services = {
    peering-manager = {
      enable = true;
      secretKeyFile = config.sops.secrets."peering-manager/secretKey".path;
      port = 8000;
      listenAddress = "127.0.0.1";
      plugins = (ps: [ (mozilla-django-oidc ps) ]);
      settings = {
        # CLIENT_ID and SECRET are required to authenticate against the provider
        OIDC_RP_CLIENT_ID = "pm.domain.de";

        # The following two may be required depending on your provider,
        # check the configuration endpoint for JWKS information
        OIDC_RP_SIGN_ALGO = "RS256";
        OIDC_OP_JWKS_ENDPOINT = "https://id.domain.de/realms/main/protocol/openid-connect/certs";

        # Refer to the configuration endpoint of your provider
        OIDC_OP_AUTHORIZATION_ENDPOINT = "https://id.domain.de/realms/main/protocol/openid-connect/auth";
        OIDC_OP_TOKEN_ENDPOINT = "https://id.domain.de/realms/main/protocol/openid-connect/token";
        OIDC_OP_USER_ENDPOINT = "https://id.domain.de/realms/main/protocol/openid-connect/userinfo";

        # Set these to the base path of your Peering Manager installation
        LOGIN_REDIRECT_URL = "https://pm.domain.de";
        LOGOUT_REDIRECT_URL = "https://pm.domain.de";

        # If this is True, new users will be created if not yet existing.
        OIDC_CREATE_USER = true;

        LOGIN_REQUIRED = true;
        TIME_ZONE = "Europe/Berlin";
        ALLOWED_HOSTS = [
          "pm.domain.de"
        ];
      };
      extraConfig = ''
        with open("${config.sops.secrets."peering-manager/oauthSecret".path}", "r") as file:
          OIDC_RP_CLIENT_SECRET = file.readline()
      '';
    };
  };
}
```

**Comment by Xyz00777** [comment]
hi i would really like to use some plugins for my netbox instance.
@minijackson any status update on your site?
sadly i dont know how to package them myself....

**Comment by minijackson** [comment]
There's currently two PRs in progress:

- https://github.com/NixOS/nixpkgs/pull/301273
- https://github.com/NixOS/nixpkgs/pull/307786

Plugins seems to be quite similar, packaging wise. I have a bunch of them currently on my working tree, but I'm discussing with my colleagues to see which one we want, so which one I'd be willing to maintain. But so far, packaging each NetBox plugin has been pretty much copying pasting.

Another I think I'll propose, is to rename `services.netbox.plugins` -> `services.netbox.extraPackages`, so that it's clearer on what it does. It may be more obvious that `settings.PLUGINS` needs to be set.

**Comment by Chaostheorie** [comment]
I've added a note about this to the [Wiki](https://wiki.nixos.org/wiki/NetBox#Plugins). On an unrelated note, could we somehow embed the plugin identifier as an attribute for `settings.PLUGINS` inside the derivations for plugins? Maybe as an extra attribute that's passed through. The module could then automatically add the plugin identifier, based on `services.netbox.plugins`, to the settings for a better UX.

**Comment by felbinger** [comment]
> I've added a note about this to the [Wiki](https://wiki.nixos.org/wiki/NetBox#Plugins). On an unrelated note, could we somehow embed the plugin identifier as an attribute for `settings.PLUGINS` inside the derivations for plugins? Maybe as an extra attribute that's passed through. The module could then automatically add the plugin identifier, based on `services.netbox.plugins`, to the settings for a better UX.

There are several plugins, that are incompatible with each other (e.g. routing + bgp, security + data flows). What do you think about adding options to the netbox module to add plugins in a declarative manner? This way we could add certain assertions to ensure the combination of plugins does actually work. Also I think it will be easier for new users to understand how to use plugins.

<details>
<summary>example of incompability between security and data flows plugins</summary>
netbox_data_flows.Application.owner: (fields.E304) Reverse accessor 'Owner.application_set' for 'netbox_data_flows.Application.owner' clashes with reverse accessor for 'netbox_security.Application.owner'.
	HINT: Add or change a related_name argument to the definition for 'netbox_data_flows.Application.owner' or 'netbox_security.Application.owner'.
netbox_data_flows.Application.tags: (fields.E304) Reverse accessor 'Tag.application_set' for 'netbox_data_flows.Application.tags' clashes with reverse accessor for 'netbox_security.Application.tags'.
	HINT: Add or change a related_name argument to the definition for 'netbox_data_flows.Application.tags' or 'netbox_security.Application.tags'.
netbox_security.Application.owner: (fields.E304) Reverse accessor 'Owner.application_set' for 'netbox_security.Application.owner' clashes with reverse accessor for 'netbox_data_flows.Application.owner'.
	HINT: Add or change a related_name argument to the definition for 'netbox_security.Application.owner' or 'netbox_data_flows.Application.owner'.
netbox_security.Application.tags: (fields.E304) Reverse accessor 'Tag.application_set' for 'netbox_security.Application.tags' clashes with reverse accessor for 'netbox_data_flows.Application.tags'.
	HINT: Add or change a related_name argument to the definition for 'netbox_security.Application.tags' or 'netbox_data_flows.Application.tags'.
</details>

**Comment by Chaostheorie** [comment]
> What do you think about adding options to the netbox module to add plugins in a declarative manner

I think the idea is neat but I'm not too sure about the maintenance side here. It would require either plugin authors or the netbox maintainers to keep somewhat up to date with the incompatibilities. This is not a particular hard task but requires attention and to some extent testing if we want to provide a reliable UX.

---

I could not find a note about your example plugins being incompatible with each other in their documentation at first glance. So presumably this would be dependent on a plugin maintainer knowing about this from experience/testing/bug reports.

If the plugins become compatible with each other in the future, then a maintainer likely won't know about it directly however. This could inconvenience a user later when they want to use both and are blocked by the assertion. This example is a bit contrived though and I'm likely overestimating the effort required and/or actual usage of this.

I personally haven't really encountered incompatible plugins, only some that just didn't work anymore on newer NetBox versions (which NetBox is addressing itself these days).

**Comment by felbinger** [comment]
You're right, interoperability between different plugins isn't being tested in the NetBox community. Assertions seam to be too strict for this. Could we develop an integration test to validate plugin compatibility? Most fatal errors happen during database migration, so checking these should catch the majority of issues.
